helix-core/src/main/java/org/apache/helix/cloud/event/helix/DefaultCloudEventCallbackImpl.java

```diff
@@ -43,27 +48,40 @@ public void disableInstance(HelixManager manager, Object eventInfo) {
    * @param eventInfo Detailed information about the event
    */
   public void enableInstance(HelixManager manager, Object eventInfo) {
-    // To be implemented
-    throw new NotImplementedException();
+    manager.getClusterManagmentTool()
+        .enableInstance(manager.getClusterName(), manager.getInstanceName(), true);
   }
 
   /**
-   *
+   * Put cluster into maintenance mode if the cluster is not currently in maintenance mode
    * @param manager The helix manager associated with the listener
    * @param eventInfo Detailed information about the event
    */
   public void enterMaintenanceMode(HelixManager manager, Object eventInfo) {
-    // To be implemented
-    throw new NotImplementedException();
+    if (!manager.getClusterManagmentTool().isInMaintenanceMode(manager.getClusterName())) {
+      manager.getClusterManagmentTool()
+          .manuallyEnableMaintenanceMode(manager.getClusterName(), true, REASON, null);
+    }
   }
 
   /**
-   *
+   * Exit maintenance mode for the cluster, if there is no more live instances disabled for cloud event
    * @param manager The helix manager associated with the listener
    * @param eventInfo Detailed information about the event
    */
   public void exitMaintenanceMode(HelixManager manager, Object eventInfo) {
```

Comment:
I am thinking about this could still be a concern for race condition. Say two instances are ready to enable, both of them check instantly and find another instance is disabled. So both of them just disabled without exiting maintenance mode.

I think we may need to introduce some centralized locking to avoid the race condition.

Reply:
As discussed offline, enable/disable instance in ClusterConfig level can also prevent race condition, and is a clearer way than adding a lock. Will update the PR to enable/disable instance through ClusterConfig once @xyuanlu made the change to re-enable batch enable/disable instances.